File: Java/src/Hard/q84.java
package Hard;

import java.util.ArrayDeque;
import java.util.Deque;

public class q84 {
    public static int largestRectangleArea(int[] heights) {
        Deque<Integer> st = new ArrayDeque<Integer>();

        int [] newHeights = new int[heights.length + 2];
        newHeights[0] = 0;
        newHeights[newHeights.length - 1] = 0;
        for (int index = 0; index < heights.length; index++){
            newHeights[index + 1] = heights[index];
        }

        heights = newHeights;

        st.push(0);
        int result = 0;
        for (int i = 1; i < heights.length; i++) {
            if (!st.isEmpty() && heights[i] == heights[st.peek()]) {
                st.pop();
            }

            while (!st.isEmpty() && heights[i] < heights[st.peek()]) {
                int mid = st.pop();

                if (!st.isEmpty()) {
                    int left = st.peek();
                    int w = i - left - 1;
                    int h = heights[mid];
                    result = Math.max(result, w * h);
                }
            }

            st.push(i);

        }
        return result;
    }

    public static void testSample() {

    }
}

File: Python/src/hard/q84.py
from typing import List
from collections import deque

class Solution:
    """
    Leetcode 84: Largest Rectangle in Histogram
    """

    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = deque([0])
        heights = [0] + heights + [0] # pad 0s for monotonic decreasing stack, and force pop at the end
        result = 0

        for i in range(len(heights)):
            if heights[i] > heights[stack[-1]]: # next > top, pushing into the stack
                stack.append(i)
            elif heights[i] == heights[stack[-1]]:
                stack.pop()
                stack.append(i)
            else: # next < top, poping and calculating max area
                while stack and heights[i] < heights[stack[-1]]:
                    mid = stack.pop()
                    right = i

                    if stack:
                        left = stack[-1]
                        w = right - left - 1 # exclude left and right, we only calculate what's in between
                        h = heights[mid]
                        result = max(result, w*h)
                stack.append(i)
        
        return result